src/lib/protocols/kerberos.c

```diff
@@ -36,6 +36,221 @@ static int ndpi_search_kerberos_extra(struct ndpi_detection_module_struct *ndpi_
+      (packet->payload[*kasn1_offset] != 0xA3 /* Context-specific Constructed Tag Type: Bit String */ &&
+       packet->payload[*kasn1_offset] != 0xA4 /* Context-specific Constructed Tag Type: Octect String */ &&
+       packet->payload[*kasn1_offset] != 0x30 /* Sequence Of */))
+  {
+    return -1;
+  }
+
+  (*kasn1_offset)++;
+
+  length = krb_decode_asn1_length(ndpi_struct, kasn1_offset);
+  if (length < 0)
+  {
+    return -1;
+  }
+
+  if (out != NULL)
+  {
+    *out = (char *)&packet->payload[*kasn1_offset];
+  }
+
+  return length;
+}
+
+static int krb_decode_asn1_blocks_skip(struct ndpi_detection_module_struct *ndpi_struct,
+                                       size_t * const kasn1_offset)
+{
+  struct ndpi_packet_struct * const packet = &ndpi_struct->packet;
+  int length;
+
+  if (packet->payload_packet_len <= *kasn1_offset + 1 /* length octet */ ||
+      (packet->payload[*kasn1_offset] != 0xA0 &&
+       packet->payload[*kasn1_offset] != 0xA1))
+  {
+    return -1;
+  }
+
+  (*kasn1_offset)++;
+
+  length = krb_decode_asn1_length(ndpi_struct, kasn1_offset);
+  if (length < 0)
+  {
+    return -1;
+  }
+
+  return length;
+}
+
+static void strncpy_lower(char * const dst, size_t dst_siz,
+                          char const * const src, size_t src_siz)
+{
+   int dst_len = ndpi_min(src_siz, dst_siz - 1);
+
+   strncpy(dst, src, dst_len);
+   dst[dst_len] = '\0';
+
+   for(int i = 0; i < dst_len; ++i)
+   {
+     dst[i] = tolower(dst[i]);
+   }
+}
+
+/* Reference: https://datatracker.ietf.org/doc/html/rfc4120 */
+static int krb_parse(struct ndpi_detection_module_struct * const ndpi_struct,
+                     struct ndpi_flow_struct * const flow,
+                     size_t kasn1_offset)
+{
+  int length;
+  char const * text;
+
+  length = krb_decode_asn1_sequence_type(ndpi_struct, &kasn1_offset); /* Optional PADATA */
+  if (length > 0)
+  {
+    kasn1_offset += length;
+  }
+
+  length = krb_decode_asn1_string_type(ndpi_struct, &kasn1_offset, &text);
+  if (length < 0)
+  {
+    return -1;
+  }
+
+  kasn1_offset += length;
+  text += 2;
+  length -= 2;
+  if (flow->protos.kerberos.domain[0] == '\0')
+  {
+    strncpy_lower(flow->protos.kerberos.domain, sizeof(flow->protos.kerberos.domain),
+                  text, length);
+  }
+
+  length = krb_decode_asn1_string_type(ndpi_struct, &kasn1_offset, NULL);
+  if (length < 0)
+  {
+    return -1;
+  }
+
+  length = krb_decode_asn1_sequence_type(ndpi_struct, &kasn1_offset);
+  if (length < 0)
+  {
+    return -1;
+  }
+
+  length = krb_decode_asn1_blocks_skip(ndpi_struct, &kasn1_offset);
+  if (length < 0)
+  {
+    return -1;
+  }
+  kasn1_offset += length;
+
+  length = krb_decode_asn1_blocks_skip(ndpi_struct, &kasn1_offset);
+  if (length < 0)
+  {
+    return -1;
+  }
+
+  length = krb_decode_asn1_string_type(ndpi_struct, &kasn1_offset, &text);
+  if (length < 0)
+  {
+    return -1;
+  }
+
+  kasn1_offset += length;
```

Comment:
Codiga warning...

Reply:
Seems like a false-positive to me. But I can change it in a way that the function parameter won't be used as a local scope variable. That should fix it.